Fix JWT Signed with Weak Algorithm

Kondent opened the pull request:
|Related issue|
|---|
|#9321|

# Description
JWT signed algorithm updated form HS256 to RS256.
Now JWT is generated with a private key and decoded with a public one. This feature gives more security against brute force attacks within a leaked JWT as discussed in related issue.
This PR closes #9321.

# Tests
## API
```
========================================================== test session starts ===========================================================
platform linux -- Python 3.9.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/kondent/git/wazuh/api
plugins: asyncio-0.15.1, html-3.1.1, metadata-1.11.0, cov-2.12.1
collected 238 items                                                                                                                      

api/models/test/test_model.py ............                                                                                         [  5%]
api/test/test_alogging.py ..........                                                                                               [  9%]
api/test/test_authentication.py ..........                                                                                         [ 13%]
api/test/test_configuration.py ..........................................                                                          [ 31%]
api/test/test_util.py ...................................                                                                          [ 45%]
api/test/test_validator.py ....................................................................................................... [ 89%]
..........................                                                                                                         [100%]

==================================================== 238 passed, 14 warnings in 1.32s ====================================================
```
## Framework 
```
==================================== test session starts =====================================
platform linux -- Python 3.9.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/kondent/git/wazuh/framework
plugins: asyncio-0.15.1, html-3.1.1, metadata-1.11.0, cov-2.12.1
collected 1589 items

wazuh/core/cluster/dapi/tests/test_dapi.py .......................                     [  1%]
wazuh/core/cluster/tests/test_cluster.py ...................                           [  2%]
wazuh/core/cluster/tests/test_common.py .......                                        [  3%]
wazuh/core/cluster/tests/test_control.py .....                                         [  3%]
...
...
...
wazuh/tests/test_stats.py ................                                             [ 94%]
wazuh/tests/test_syscheck.py .......................                                   [ 95%]
wazuh/tests/test_syscollector.py ............                                          [ 96%]
wazuh/tests/test_task.py ............................                                  [ 98%]
wazuh/tests/test_vulnerability.py ..........................                           [100%]

======================= 1589 passed, 14 warnings in 201.16s (0:03:21) ========================
```

mcarmona99 commented on api/api/test/test_authentication.py:
```diff
@@ -72,44 +72,45 @@ def test_check_user(mock_raise_if_exc, mock_submit, mock_distribute_function, mo
 
 
-@patch('api.authentication.token_urlsafe', return_value='test_token')
+@patch('api.authentication.change_keyPair', return_value=["b'-----BEGIN PRIVATE KEY-----\n'", "b'-----BEGIN PUBLIC KEY-----\n'"])
```
Not a real use case. `change_keyPair` will return bytes and no strings. Review this test as it will change.

mcarmona99 commented on api/api/authentication.py:
```diff
@@ -82,36 +85,52 @@ def check_user(user, password, required_scopes=None):
+        encoding = serialization.Encoding.PEM,
+        format = serialization.PublicFormat.SubjectPublicKeyInfo
+    )
+    with open(_private_key_path, 'wb') as keyFile:
+        keyFile.write(privateKey)
+    with open(_public_key_path, 'wb') as keyFile:
+        keyFile.write(publicKey)
```
Why are we writing bytes but reading without the 'b' flag?

mcarmona99 commented on api/api/test/test_authentication.py:
```diff
@@ -72,44 +72,45 @@ def test_check_user(mock_raise_if_exc, mock_submit, mock_distribute_function, mo
+    result = authentication.generate_keyPair()
+    assert isinstance(result[0], str)
+    assert isinstance(result[1], str)
+    assert result == ["b'-----BEGIN PRIVATE KEY-----\n'", "b'-----BEGIN PUBLIC KEY-----\n'"]
```
The result will be `["---....", "-----....."]` (without 'b') as we are reading without the `b` flag. This will change depending on the decision taken in one of the last comments where this was commented.

The tests that follow will need to be changed too.

Kondent commented on api/api/authentication.py:
```diff
@@ -6,6 +6,9 @@
+from cryptography.hazmat.primitives import serialization
+
 from concurrent.futures import ThreadPoolExecutor
 from secrets import token_urlsafe
 from shutil import chown
```
Unused imports removed.

Kondent commented on api/api/authentication.py:
```diff
@@ -6,6 +6,9 @@
 import copy
 import logging
 import os
+from cryptography.hazmat.primitives.asymmetric import rsa                                 
+from cryptography.hazmat.primitives import serialization
+
```
Imports order fixed.

Kondent commented on api/api/authentication.py:
```diff
@@ -82,36 +85,52 @@ def check_user(user, password, required_scopes=None):
 
-def generate_secret():
-    """Generate secret file to keep safe or load existing secret."""
+def generate_keyPair():
```
Function and variable names changed, following PEP8 standard.

Kondent commented on api/api/authentication.py:
```diff
@@ -82,36 +85,52 @@ def check_user(user, password, required_scopes=None):
+    return [privateKey, publicKey]
+
+
+def change_keyPair():
```
PEP8 standard related warnings were fixed.

Kondent commented:
Hi
I just fixed pep8 standard warnings from all files, how keys are saved and recovered from files.
Added consistency in data types: now keys are stored as strings and jwt encode/decode methods use them as strings too.
I ran tests again just to be sure everything is working properly after this fix.

Regards,
Alexis

mcarmona99 commented on api/api/test/test_authentication.py:
```diff
@@ -72,44 +72,47 @@ def test_check_user(mock_raise_if_exc, mock_submit, mock_distribute_function, mo
-        authentication.generate_secret()
-
-        calls.append(call(authentication._secret_file_path, mode='r'))
+        calls = [call(authentication._private_key_path, mode='rb'), call(authentication._public_key_path, mode='rb')]
         mock_open.has_calls(calls)
```
There seems to be something wrong. This test should not pass as the `open` method doesn't have these calls (we need another call to `authentication.generate_keypair()` (inside the context of `os.path.exists` with `return_value=True`).

In case we had the `authentication.generate_keypair()`, the test should still fail as the calls are not those ones.

mcarmona99 commented on api/api/test/test_authentication.py:
```diff
@@ -72,44 +72,47 @@ def test_check_user(mock_raise_if_exc, mock_submit, mock_distribute_function, mo
+    result = authentication.change_keypair()
+    assert isinstance(result[0], str)
+    assert isinstance(result[1], str)
+    calls = [call(authentication._private_key_path, mode='wb'), call(authentication._public_key_path, mode='wb')]
+    mock_open.has_calls(calls)
```
Similar to the comment above, calls are not the right ones. Mode is w and the test is passing with wb.

mcarmona99 reviewed (dismissed):
LGTM!

Kondent commented:
## PR Update:
RS256 algorithm changed to ES512 (ECDSA signature algorithm using SHA-512).

* Advantages:
  * Shorter keys than RS256.
  * Shorter JWT.
  * Higher security level.
